modules/currency-l0/src/main/scala/io/constellationnetwork/currency/l0/snapshot/services/StateChannelBinarySender.scala

```diff
@@ -67,168 +72,212 @@ object State {
+          case PendingBinary(_, enqueuedAtOrdinal, _) =>
+            currentOrdinal.value - enqueuedAtOrdinal.value >= noConfirmationsToTriggerRetryMode
+          case _ => false
+        }
+
+        val updatedRetryMode = if (!state.retryMode) {
+          hasStalled
+        } else {
+          val pendingCount = state.tracked.collect { case _: PendingBinary => 1 }.sum
+          val allPendingAlreadySent = state.tracked.forall {
+            case PendingBinary(_, _, sendsSoFar) => sendsSoFar >= 1
+            case _                               => true
+          }
+
+          if (pendingCount <= state.cap && allPendingAlreadySent && !hasStalled)
+            false
+          else
+            true
+        }
+
+        state.copy(retryMode = updatedRetryMode)
+      }
+
+      private def updateRetryParameters(
+        state: State,
+        previousRetryMode: Boolean
+      ): State =
+        if ((!state.retryMode && previousRetryMode) || state.tracked.isEmpty) {
+          State.empty.copy(tracked = state.tracked)
+        } else if (!state.retryMode) {
+          state
+        } else {
+          val confirmedCount = state.tracked.count(_.isInstanceOf[ConfirmedBinary])
+
+          if (confirmedCount > 0) {
+            val maxCap = confirmedCount * confirmedCountMultiplier
+            val log2 = (x: Double) => Math.log10(x) / Math.log10(2.0)
+            val surplus = NonNegLong.from {
+              Math.ceil(log2(state.tracked.length.toDouble)).toLong
+            }.getOrElse(NonNegLong.MinValue)
+            val proposedCap = state.cap + surplus
+            val updatedCap = NonNegLong.from(Math.min(proposedCap, maxCap)).getOrElse(NonNegLong.MinValue)
+
+            state.copy(
+              cap = updatedCap,
+              backoffExponent = NonNegLong(0L),
+              noConfirmationsSinceRetryCount = NonNegLong(0L)
+            )
+          } else if (state.cap > 1) {
+            state.copy(
+              cap = NonNegLong.from(state.cap - 1).getOrElse(NonNegLong.MinValue),
+              backoffExponent = NonNegLong(0L),
+              noConfirmationsSinceRetryCount = NonNegLong(0L)
+            )
+          } else if (state.cap.value == 1) {
+            state.copy(
+              cap = NonNegLong(0L),
+              backoffExponent = NonNegLong.from(state.backoffExponent + 1L).getOrElse(NonNegLong.MaxValue),
+              noConfirmationsSinceRetryCount = NonNegLong(1L)
+            )
+          } else {
+            val noConfirmationsSinceRetryCount =
+              NonNegLong.from(state.noConfirmationsSinceRetryCount + 1).getOrElse(NonNegLong.MaxValue)
+            val updatedCap =
+              if (noConfirmationsSinceRetryCount.toLong >= Math.ceil(Math.pow(2.0, state.backoffExponent.toDouble)).toLong)
+                NonNegLong(1L)
+              else state.cap
+
+            state.copy(
+              cap = updatedCap,
+              noConfirmationsSinceRetryCount = noConfirmationsSinceRetryCount
```

Comment:
Most of values in this function would be good to track as metrics to see if we're hitting a retry logical fault. (As in the retry wait delay is too long.)